### PR TITLE
members only Authentication

### PR DESCRIPTION
**Pull request by Amalcxc**
**we builded an exclusive clubhouse where your members can write anonymous posts.**

1. Createdyour new members-only Rails app
2. Created a Post model and a Posts controller
3. used a #before_action to restrict access
5. created form view
6. add logic in your Index view to display the author’s name

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,106 @@
...
+- Web browser other than Internet Explorer
 
-* How to run the test suite
+### Setup
 
-* Services (job queues, cache servers, search engines, etc.)
+- open terminal in your computer
+- run 'git clone https://github.com/Anvi98/member-only.git'
+- cd member-only
+- type "bundler install"
```
Comment (Dipeshtwis) [inline README.md:23]: After bundle install. There should be rails migration command.

Hint: `rails db:migrate`

File: app/views/posts/index.html.erb [hunk 2]
```diff
@@ -16,14 +21,32 @@
...
+<% if user_signed_in? %>
+   <p>
+     <%= link_to “Logout”, destroy_user_session_path, method: :delete %>
+   </p>
+ <% else %>
+ <p class=“control”>
+   <%= link_to ‘Sign In’, new_user_session_path%>
+ </p>
+ <p class=“control”>
+   <%= link_to ‘Sign Up’, new_user_registration_path%>
```
Comment (Dipeshtwis) [inline app/views/posts/index.html.erb]: Instead **‘Sign Up’**. It should be 'Sign Up'. See the `Apostrophe`.

File: app/views/posts/index.html.erb [hunk 3]
```diff
@@ -16,14 +21,32 @@
...
     <% end %>
   </tbody>
 </table>
 
 <br>
 
-<%= link_to 'New Post', new_post_path %>
+<% if user_signed_in? %>
+   <p>
+     <%= link_to “Logout”, destroy_user_session_path, method: :delete %>
```
Comment (Dipeshtwis) [inline app/views/posts/index.html.erb]: Instead **“Logout”**. It should be **'Logout'**.

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -1,32 +1,30 @@
 class PostsController < ApplicationController
   before_action :set_post, only: %i[show edit update destroy]
-  before_action :authenticate_user!, except: %i[new create]
 
   # GET /posts or /posts.json
```
Comment (Dipeshtwis) [inline app/controllers/posts_controller.rb]: Remove all the commented code.

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -1,32 +1,30 @@
...
   end
 
   # GET /posts/1 or /posts/1.json
-  def show
-  end
+  def show; end
 
   # GET /posts/new
   def new
     @post = current_user.posts.build
```
Comment (Dipeshtwis) [inline app/controllers/posts_controller.rb]: Post initializes an empty post `@post = Post.new`. As we are restricting users with before_action for new and create method.

File: app/views/posts/index.html.erb [hunk 6]
```diff
@@ -16,14 +21,32 @@
...
       </tr>
     <% end %>
   </tbody>
 </table>
 
 <br>
 
-<%= link_to 'New Post', new_post_path %>
+<% if user_signed_in? %>
+   <p>
```
Comment (Dipeshtwis) [inline app/views/posts/index.html.erb:42]: This whole thing is in `application.html.erb` ( signin/signup). so that it will be available for all the views.

**Review by Dipeshtwis — CHANGES_REQUESTED** [review]
# Status:  ♻️ Changes Required ♻️
 Hi @Amalcxc  and @Anvi98 

**You've a;most hit all the requirements in one shot. Well done.** Here're the major highlights of this PR:
- Your PR title and description are really descriptive.
- Used Devise gem to authenticate the user.

**However, in order to approve, we have to work a little bit more. Please check my comments below and fix them as appropriate.**

## For Approval

- [x] I can't able to create a post after login as there is no button for creating a post. I had to go by writing `posts/new` to create a new post. Please specify the button in view which takes me to the post creation page.
- [x] The application.html.erb has Sign in/Sign out links  so that it will present in any view of the app ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_out_link_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/)).

- [x] Remove all the commented code.

- [x] Post controller should use `before_action` to restrict access to the `:new` and `:create` methods (uses ApplicationController helper method) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).
- [x] Your README is missing one step in installing. I have mentioned in the in-line comment, please have a look.

## Please make the required changes and submit another review request when you are done.

(Feel free to contact me on slack by the handle of  **deep** if you have any additional questions).
Happy Coding
[Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️
Hey @Amalcxc and @Anvi98, You have done a good job Implementing almost all the changes requested by the previous reviewer, However, there is one issue that has prevented this project from being approved.

- [x] I can't log out because I can't see a log out button, please add a log out button and make another submission.

In case you have a question, you can reach me on slack as Elisha kyakopo.

**Review by jubaan — APPROVED** [review]

**Review by jubaan — APPROVED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> 
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🥳 🎊 🎉 🥳  APPROVED 🥳 🎊 🎉 🥳 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉
</h1>
<h2 align="center">
  What’s done is done. What’s gone is gone. <br />
  One of life’s lessons is always moving on. <br />
  It’s okay to look back to see how far you’ve come but <br /> 
  KEEP MOVING FORWARD.
</h2>

<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>


<h3 id="optional"> 1. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] If you have something here, even when approved, it's just a little push to improve your project.</h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 2. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 3. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<p> How does it feel? Is nice to advance in the program, right? It keeps you motivated even with all the struggles and takes you one step closer to the goal 🎯!. I wan't you to keep advancing so if you feel stuck and/or need help, remember there is always someone you can reach to keep moving forward, doesn't mater how "dum" you think the questions is.</p>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h1 align="center"> See you a the top 🚀!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Pull request merged**